packages/@aws-cdk/aws-codepipeline-actions/lib/action.ts

```diff
@@ -45,6 +67,22 @@ export abstract class Action implements codepipeline.IAction {
     return rule;
   }
 
+  protected variableWasReferenced(): void {
+    this.variableReferenced = true;
+  }
+
+  protected variableExpression(variableName: string): string {
+    return Lazy.stringValue({ produce: () => {
+      // make sure the action was bound (= added to a pipeline)
+      if (this.actualNamespace) {
+        return `#{${this.actualNamespace}.${variableName}}`;
+      } else {
+        throw new Error(`Cannot reference variables of action '${this.actionProperties.actionName}', ` +
+          'as that action was never added to a pipeline');
+      }
```

Comment:
Instead of making the variable expression a lazy value, just make the actual namespace a lazy value. This way you can get rid of the error-prone `variableWasReferenced` method:

```ts
protected variableExpression(variableName: string): string {
  this.variableReferenced = true;
  return `#{${this.variablesNamespace}}.${variableName}`;
}
```

And then in the initializer:

```ts
this.variablesNamespace = actionProperties.variablesNamespace 
  ?? Lazy.string({ produce: () => this.renderedNamespace });
```

Reply:
Isn't this API even more error-prone? If you decide to cache the variables, like I did originally:

```ts
export class GitHubSourceAction extends Action {
  constructor(props: GitHubSourceActionProps) {
    // ...
    this._variables = {
      repositoryName: this.variableExpression('RepositoryName'),
      branchName: this.variableExpression('BranchName'),
      authorDate: this.variableExpression('AuthorDate'),
      committerDate: this.variableExpression('CommitterDate'),
      commitId: this.variableExpression('CommitId'),
      commitMessage: this.variableExpression('CommitMessage'),
      commitUrl: this.variableExpression('CommitUrl'),
    };
  }
```

, instead of returning them on demand from the `variables` getter, you will always make `variableReferenced` `true`, even if no variables were actually referenced, and thus always render the namespace, even if it wasn't required.